### PR TITLE
Uriel/hwvis 792

### PR DESCRIPTION
**Pull request by UrielHaile**
PR for https://aeolusbot.atlassian.net/browse/HWVIS-792

A discussion here is how to properly close the driver.

### CODE REVIEW

File: openni2_camera/src/openni2_device.cpp [hunk 1]
```diff
@@ -682,8 +682,14 @@ boost::shared_ptr<openni::VideoStream> OpenNI2Device::getIRVideoStream() const n
       ir_video_stream_ = boost::make_shared<openni::VideoStream>();
 
       const openni::Status rc = ir_video_stream_->create(*openni_device_, openni::SENSOR_IR);
-      if (rc != openni::STATUS_OK)
-        THROW_OPENNI_EXCEPTION("Couldn't create IR video stream: \n%s\n", openni::OpenNI::getExtendedError());
+      if (rc != openni::STATUS_OK) {
+        THROW_OPENNI_EXCEPTION("Couldn't create IR video stream: \n%s\n",
+                               openni::OpenNI::getExtendedError());
+
+        stopAllStreams();
```
Comment (abrdej) [inline openni2_camera/src/openni2_device.cpp]: I'm not sure about this. Does it mean that if we close the driver our reconnection handling will restart the driver properly?

Comment (abrdej) [reply openni2_camera/src/openni2_device.cpp]: One more! If we throw an exception then code after that will not be reached

Comment (UrielHaile) [reply openni2_camera/src/openni2_device.cpp]: I had tried with ros::requestShutdown(), it stops the node, that means no reconnection

File: openni2_camera/src/openni2_device.cpp [hunk 2]
```diff
@@ -43,6 +43,7 @@
 #include <boost/shared_ptr.hpp>
 #include <boost/make_shared.hpp>
 
+#include <ros/ros.h>
```
Comment (abrdej) [inline openni2_camera/src/openni2_device.cpp]: for `ros::requestShutdown()` include just `#include <ros/init.h>`

File: openni2_camera/src/openni2_device.cpp [hunk 3]
```diff
@@ -680,10 +681,11 @@ boost::shared_ptr<openni::VideoStream> OpenNI2Device::getIRVideoStream() const n
     if (hasIRSensor())
     {
       ir_video_stream_ = boost::make_shared<openni::VideoStream>();
-
       const openni::Status rc = ir_video_stream_->create(*openni_device_, openni::SENSOR_IR);
-      if (rc != openni::STATUS_OK)
-        THROW_OPENNI_EXCEPTION("Couldn't create IR video stream: \n%s\n", openni::OpenNI::getExtendedError());
+      if (rc != openni::STATUS_OK) { 
+        ROS_ERROR("Couldn't create IR video stream: \n%s\n", openni::OpenNI::getExtendedError());
+        ros::requestShutdown();
```
[the remark this answers is not in the document]
Comment (abrdej) [reply openni2_camera/src/openni2_device.cpp]: I found that `ros::requestShutdown()` will close the whole nodelet manager - we can't do this as inside will be almost all our code.
After small tests it even not restart properly manager and nodelets inside.

Seems that we need to find another solution.
Maybe nodelet load/unload service can help but I thing that the better idea is to find a way how to handle some openni errors without restarting nodelet. It should be possible just by closing all the things that can block the device and restart them.

[the remark this answers is not in the document]
Comment (abrdej) [reply openni2_camera/src/openni2_device.cpp]: Also if we handle this case for IR stream - why not for depth (and even for RGB)? :)

File: openni2_camera/src/openni2_device.cpp (hunk 2 above)
Comment (UrielHaile) [reply openni2_camera/src/openni2_device.cpp]: got it, thanks

**Pull request closed without merging**